src/sentry/integrations/bitbucket_server/integration.py

```diff
@@ -312,16 +326,40 @@ def get_unmigratable_repositories(self):
         return list(filter(lambda repo: repo.name not in accessible_repos, repos))
 
     def source_url_matches(self, url: str) -> bool:
-        raise IntegrationFeatureNotImplementedError
+        return url.startswith(self.model.metadata["base_url"])
 
     def format_source_url(self, repo: Repository, filepath: str, branch: str | None) -> str:
-        raise IntegrationFeatureNotImplementedError
+        project = quote(repo.config["project"])
+        repo_name = quote(repo.config["repo"])
+        source_url = f"{self.model.metadata["base_url"]}/projects/{project}/repos/{repo_name}/browse/{filepath}"
+
+        if branch:
+            source_url += "?" + urlencode({"at": branch})
+
+        return source_url
 
     def extract_branch_from_source_url(self, repo: Repository, url: str) -> str:
-        raise IntegrationFeatureNotImplementedError
+        parsed_url = urlparse(url)
+        qs = parse_qs(parsed_url.query)
+
+        if "at" in qs and len(qs["at"]) == 1:
+            branch = qs["at"][0]
+
+            # branch name may be prefixed with refs/heads/, so we strip that
+            refs_prefix = "refs/heads/"
+            if branch.startswith(refs_prefix):
+                branch = branch[len(refs_prefix) :]
+
+            return branch
+
+        return ""
 
     def extract_source_path_from_source_url(self, repo: Repository, url: str) -> str:
-        raise IntegrationFeatureNotImplementedError
+        if repo.url is None:
+            return ""
+        parsed_repo_url = urlparse(repo.url)
+        parsed_url = urlparse(url)
+        return parsed_url.path.replace(parsed_repo_url.path + "/", "")
```

Comment:
i'm also curious how this is different from bitbucket, it kind of looks the same?